pkg/i2gw/provider.go

```diff
@@ -73,6 +73,16 @@ type ResourceConverter interface {
 	ToGatewayAPI(resources InputResources) (GatewayResources, field.ErrorList)
 }
 
+// HTTPPathMatchOption is an option to customize the ingress implementationSpecific
+// match type conversion.
+type HTTPPathMatchOption func(*gatewayv1beta1.HTTPPathMatch)
+
+// ImplementationSpecificOptions contains all the pointers to implementation-specific
+// customization functions.
+type ImplementationSpecificOptions struct {
+	HTTPPathmatch HTTPPathMatchOption
+}
```

Comment:
Can you explain the idea behind the names?

What would be the use of HttpMatchOption? only for implementation specific funcs or there is a vision it could be used for other things? 

Similarly for `ImplementationSpecificOptions` - is this intended to hold various different things/options that the implementation would need during the conversion? If yes, I would prefer be consistent and calling it "provider" and not implementation. Do shout if I got it wrong.

Reply:
Another question, @levikobi feel free to jump here as well. 

Why do we need another struct to hold provider specific options if we have `ProviderConf` cant we just initialize providerConf with the relevant `HTTPPathMatchOption` and pass `c.conf` to [ToGateway](https://github.com/kubernetes-sigs/ingress2gateway/blob/main/pkg/i2gw/providers/kong/converter.go#L50) here instead of the new `ImplementationSpecificOptions` struct?

Reply:
> What would be the use of HttpMatchOption? only for implementation specific funcs or there is a vision it could be used for other things?

it's intended to be used only on the `implementationSpecific` path type. Probably the name is a bit unfortunate, I'll rename to reflect that 👍 .
All the handlers in `ImplementationSpecificOptions` should be intended to customize the `implementationSpecific` fields of the API. I'll add a comment about that.

> is this intended to hold various different things/options that the implementation would need during the conversion? If yes, I would prefer be consistent and calling it "provider" and not implementation

it's indented to hold all the handlers to customize the `Implementation-specific` fields of the APIs. `Implementation-specific` is the term commonly used in the Gateway API and I would rely on it. Maybe we can rename it `ProviderImplementationSpecificOptions`. WDYT?

> Why do we need another struct to hold provider specific options if we have ProviderConf cant we just initialize providerConf with the relevant HTTPPathMatchOption and pass c.conf to [ToGateway](https://github.com/kubernetes-sigs/ingress2gateway/blob/main/pkg/i2gw/providers/kong/converter.go#L50) here instead of the new ImplementationSpecificOptions struct?

The problem with the ProviderConf is that it is instantiated outside the provider and given to the provider. As it is designed at the moment, it's a good way to pass parameters from outside to the provider (such as the client). The other way around results to be a bit clunky at the moment. The implementation-specific function handlers are intended to be passed from the provider to the common package, and relying on the `ProviderConf` to do so is a bit odd as it is designed today in my opinion.

Reply:
> it's indented to hold all the handlers to customize the Implementation-specific fields of the APIs. Implementation-specific is the term commonly used in the Gateway API and I would rely on it. Maybe we can rename it ProviderImplementationSpecificOptions. WDYT?

so when you say `implementationSpecific` fields of the API you basically mean "whatever custom thing the implementation will need during the conversion", right? 
If so the project convention is "Provider" instead of implementation. I agree with you we need to be consistent with Gateway API, but this requires further discussion. We better be consistent within the project now, and if we decide to change it we should do the wording change in one PR.
Provider == implementation in this case so I think we need something else than `ProviderImplementationSpecificOptions`. (I think we can just add the `implementationSpecific` function to the ProviderConf`

 

> The problem with the ProviderConf is that it is instantiated outside the provider and given to the provider. As it is designed at the moment, it's a good way to pass parameters from outside to the provider (such as the client). The other way around results to be a bit clunky at the moment. The implementation-specific function handlers are intended to be passed from the provider to the common package, and relying on the ProviderConf to do so is a bit odd as it is designed today in my opinion.

cant you edit the conf [here](https://github.com/kubernetes-sigs/ingress2gateway/blob/main/pkg/i2gw/providers/kong/kong.go#L39C30-L45) and add the `implementationSpecificHTTPPathMatch` function field. Then call new provider with the new updated conf.

I think the idea behind initializing the provider conf in the common package is to provide implementation with the default client, but they can override it. I think `implementationSpecificHTTPPathMatch` should be a common func field to every implementation hence I think it should be in `ProviderConf`

Reply:
> so when you say implementationSpecific fields of the API you basically mean "whatever custom thing the implementation will need during the conversion", right?

when I say implementation-specific, I refer to the Gateway API support levels [terminology](https://gateway-api.sigs.k8s.io/concepts/conformance/#2-support-levels) (core, extended, implementation-specific). Since core and extended fields can be converted in the common package, as they are implementation-independent and part of the API, all the implementation-specific features, such as the regex path type for HTTPRoute, require a custom modifier to be set. For this reason, I'd keep the `ImplementationSpecific` part of the options, as they refer to the actual terminology. I set `ProviderImplementationSpecificOptions` as these are options related to `implementationSpecific` settings of the provider.

> cant you edit the conf [here](https://github.com/kubernetes-sigs/ingress2gateway/blob/main/pkg/i2gw/providers/kong/kong.go#L39C30-L45) and add the implementationSpecificHTTPPathMatch function field ? Then call new provider with the new updated conf.

This is the clunkiness I was talking about. I don't like the approach of putting everything into the config before creating the provider, but It's a personal opinion, it works either way.

Reply:
And what if we put such handlers into the converter, along with the `featureParsers`? It looks like the very best place to put such functions: we do not need to change the conf passed from common and can keep such field under the provider's control in a structured way. WDYT?

Reply:
Nice catch! I agree that `featureParser` is the best way to put it. with this featureParser implemented I will suggest as well to change https://github.com/kubernetes-sigs/ingress2gateway/blob/main/pkg/i2gw/providers/common/converter.go#L364 to print different warning message like "implementationSpecific path type is not supported in generic translation, however providers may added a custom support to translate it. Check your HttpRoute/s to see they reflect the implementationSpecific path type"

Dont take this exact wording, I was just mumbling without thinking too much :)

Reply:
> And what if we put such handlers into the converter, along with the `featureParsers`? It looks like the very best place to put such functions: we do not need to change the conf passed from common and can keep such field under the provider's control in a structured way. WDYT?

+1
This is exactly what I wanted

Reply:
👍 options moved to the converter